apps/aefate/src/aefa_encoding.erl

```diff
@@ -0,0 +1,349 @@
+    end;
+serialize(Map) when ?IS_FATE_MAP(Map) ->
+    L = [{_K,_V}|_] = maps:to_list(?FATE_MAP_VALUE(Map)),
+    Size = length(L),
+    %% TODO:  check all K same type, and all V same type
+    %%        check K =/= map
+    Elements = << <<(serialize(K1))/binary, (serialize(V1))/binary>> || {K1,V1} <- L >>,
+    <<?MAP,
+      (rlp_integer(Size))/binary,
+      (Elements)/binary>>;
+%% serialize(?FATE_VARIANT(Tag, Values)) when ?IS_FATE_INTEGER(Tag),
+%%                                            ?IS_FATE_TUPLE(Values) ->
+serialize(?FATE_VARIANT(Tag, Values)) when 0 =< Tag
+                                           , Tag < 256 ->
+    <<?VARIANT,
+      (serialize(?MAKE_FATE_INTEGER(Tag)))/binary,
+      (serialize(?FATE_TUPLE(Values)))/binary
+    >>.
+
+
+%% -----------------------------------------------------
+
+
+value_to_typerep(?FATE_UNIT)                 -> <<0>>;
+value_to_typerep(?FATE_TRUE)                 -> <<2>>;
+value_to_typerep(?FATE_FALSE)                -> <<2>>;
+value_to_typerep(I) when ?IS_FATE_INTEGER(I) -> <<1>>;
+value_to_typerep(A) when ?IS_FATE_ADDRESS(A) -> <<3>>;
+value_to_typerep(S) when ?IS_FATE_STRING(S)  -> <<4>>;
+value_to_typerep(T) when ?IS_FATE_TUPLE(T)   ->
+    ?FATE_TUPLE(Tuple) = T,
+    S = size(Tuple),
+    Es = << <<(value_to_typerep(E))/binary>> || E <- tuple_to_list(Tuple) >>,
+    <<5, S, Es/binary>>;
+value_to_typerep(?FATE_NIL) -> <<6>>;
+value_to_typerep(L) when ?IS_FATE_LIST(L) ->
+    [E|_] = ?FATE_LIST_VALUE(L),
+    <<7, (value_to_typerep(E))/binary >>;
+value_to_typerep(M) when ?IS_FATE_MAP(M) ->
+    Map = ?FATE_MAP_VALUE(M),
+    case maps:to_list(Map) of
+        [{A,B}|_] ->
+            <<8,
+              (value_to_typerep(A))/binary,
+              (value_to_typerep(B))/binary>>;
+        [] -> <<10>>
+    end;
+value_to_typerep(?FATE_VARIANT(Tag, T)) when Tag < 256 ->
+    L = tuple_to_list(T),
+    Size = size(T),
+    <<9,
+      (value_to_typerep(?MAKE_FATE_INTEGER(Tag)))/binary,
+      (value_to_typerep(?MAKE_FATE_INTEGER(Size)))/binary,
+      (<< << (value_to_typerep(Type))/binary >> || Type <- L >>)/binary
+    >>.
+
+
+type_to_typerep(unit)        -> <<0>>;
+type_to_typerep(integer)     -> <<1>>;
+type_to_typerep(boolean)     -> <<2>>;
+type_to_typerep(address)     -> <<3>>;
+type_to_typerep(string)      -> <<4>>;
+type_to_typerep(empty_tuple) -> <<5, 0>>;
+type_to_typerep({tuple, T}) ->
+    S = size(T),
+    Es = << <<(type_to_typerep(E))/binary>> || E <- tuple_to_list(T) >>,
+    <<5, S, Es/binary>>;
+type_to_typerep(nil)         -> <<6>>;
+type_to_typerep({list, T})   -> <<7, (type_to_typerep(T))/binary >>;
+type_to_typerep(empty_map)   -> <<10>>;
+type_to_typerep({map, A, B}) -> <<8, (type_to_typerep(A))/binary, (type_to_typerep(B))/binary >>;
+type_to_typerep({variant, Cases}) ->
+    <<9, (length(Cases)),
+      <<
+        << (length(Types)),
+           << << (type_to_typerep(Type))/binary >> || Type <- Types >>
+        >>
+        || Types <- Cases >>
+    >>.
```

Comment:
We're going to need typereps to be first class values, so I don't think it makes sense to have this function (and `deserialize_typerep` below) separate from `serialize` and `deserialize`.

Reply:
Agree. I will rework the type reps and see how to worke them in as first class values and if we should have typereps for variant types always together with their values.

Reply:
Removed typreps for now.